cmd/terramate/cli/run_script.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2023 Terramate GmbH
+// SPDX-License-Identifier: MPL-2.0
+
+package cli
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"os/exec"
+
+	"github.com/fatih/color"
+	"github.com/rs/zerolog/log"
+	"github.com/terramate-io/terramate/config"
+	"github.com/terramate-io/terramate/globals"
+	"github.com/terramate-io/terramate/hcl/eval"
+	prj "github.com/terramate-io/terramate/project"
+	"github.com/terramate-io/terramate/run"
+	"github.com/terramate-io/terramate/stdlib"
+)
+
+func (c *cli) runScript() {
+	logger := log.With().
+		Str("action", "cli.runScript()").
+		Str("workingDir", c.wd()).
+		Logger()
+
+	c.gitSafeguardDefaultBranchIsReachable()
+	c.checkOutdatedGeneratedCode()
+
+	var stacks config.List[*config.SortableStack]
+	if c.parsedArgs.Experimental.Script.Run.NoRecursive {
+		st, found, err := config.TryLoadStack(c.cfg(), prj.PrjAbsPath(c.rootdir(), c.wd()))
+		if err != nil {
+			logger.Fatal().Msgf("failed to load stack in current directory: %s", err)
+		}
+
+		if !found {
+			logger.Fatal().Msg("--no-recursive provided but no stack found in the current directory")
+		}
+
+		stacks = append(stacks, st.Sortable())
+	} else {
+		var err error
+		stacks, err = c.computeSelectedStacks(true)
+		if err != nil {
+			logger.Fatal().Msgf("failed to compute selected stacks %s", err)
+		}
+	}
+
+	// search for the script and prepare a list of script/stack entries
+	m := newScriptsMatcher(c.parsedArgs.Experimental.Script.Run.Labels)
+	m.Search(c.cfg().Tree(), stacks)
+
+	if len(m.Results) == 0 {
+		c.output.MsgStdErr(color.RedString("script not found: ") +
+			strings.Join(c.parsedArgs.Experimental.Script.Run.Labels, " "))
+		os.Exit(1)
+	}
+
+	for _, result := range m.Results {
+		c.output.MsgStdErr("Found %s defined at %s having %s job(s)",
+			color.GreenString(result.ScriptCfg.Name()),
+			color.BlueString(result.ScriptCfg.Range.String()),
+			color.BlueString(fmt.Sprintf("%d", len(result.ScriptCfg.Jobs))),
+		)
+
+		if c.parsedArgs.Experimental.Script.Run.DryRun {
+			c.output.MsgStdErr("This is a dry run, commands will not be executed.")
+		}
+
+		if len(result.Stacks) == 0 {
+			c.output.MsgStdOut(color.RedString("(skipping) ") +
+				"provided criteria does not allow script execution",
```

Comment:
Example: if you have 1000 stacks and then you chdir into a leaf stack and execute `tm script run ...` there, it will show this "(skipping) ..." message 999 times.

Reply:
I agree the message could be useful in some cases, and useless/misleading in others. Maybe only show a message when no stacks were selected at all?

Reply:
@i4ki @snakster I think we can drop the message for both cases as it feels redundant - it's clear from the rest of the output which stacks were selected.

Reply:
I decided to take it out for now and force pushed the changes, but I am happy to tweak the behaviour of that message in the next PR if needed.